Kadai1 mizushima

MizushimaToshihiko opened the pull request:
課題1を作成致しました。
どなたかレビューいただけますと幸いです。
よろしくお願いいたします。

MizushimaToshihiko committed:
first commit

MizushimaToshihiko committed:
.gitignore修正

MizushimaToshihiko commented:
## 画像変換コマンド

### 下記の仕様を満たす
- ディレクトリを指定
- 指定したディレクトリ以下のJPGファイルをPNGに変換（デフォルト）
- ディレクトリ以下は再帰的に処理する
- 変換前と変換後の画像形式を指定できる（オプション）


### 下記を満たすように開発
- mainパッケージと分離する
- 自作パッケージと標準パッケージと準標準パッケージのみ使う
- 標準標準パッケージ：golang.org/x以下のパッケージ
- ユーザ定義型を作ってみる
- GoDocを生成してみる
- Go Modulesを使ってみる

### コマンドラインオプション

 | オプション | 説明 | デフォルト |
 | --- | --- | --- |
 | -pre | 変換前フォーマット | jpeg |
 | -post | 変換後フォーマット | png |


### 使い方
- バイナリビルド（実行ファイル作成）
```bash
$ make
```
- バイナリビルド & テスト
```bash
$ make test
```

MizushimaToshihiko committed:
no new line at end of fileエラー修正

MizushimaToshihiko committed:
main.go,pic-conv/picconvert.goのリファクタリング
go fmtの適用
処理完了後のメッセージ出力

MizushimaToshihiko committed:
README.md変更、Makefileのcleanを変更（実行ファイルも削除)

MizushimaToshihiko committed:
README.mdに対応フォーマット追記

gosagawa commented on kadai1/Mizushima/main.go:
```diff
@@ -0,0 +1,57 @@
+	flag.Parse()
+}
+
+func isCorresponds(format string) bool {
```
correspondは一致するというような意味合いなので、引数を二つ渡すか、何かの型のレシーバとしての関数でないと何と一致するかが抜けて落ちているので関数名として違和感があります。
isSupportedFormat, isValidFormat、isEnableExtentionとかが適切な関数名かとおもいます。

MizushimaToshihiko replied:
お忙しいところレビュー頂きありがとうございます。
isSupportedFormatに変更させていただきました。

gosagawa commented on kadai1/Mizushima/main.go:
```diff
@@ -0,0 +1,57 @@
+
+	for _, e := range ext {
+		if e == format {
+			flag = true
```
この時点でreturn  true, ループを抜けたらreturn falseでflagという変数を使わなくても良さそうです。もしflagを使いたいとしてもループを抜けたらreturn flagで良さそうです(後のif文が入らなそうです)

MizushimaToshihiko replied:
変な処理でしたね。ありがとうございます。

gosagawa commented on kadai1/Mizushima/main.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+// errHandler occur errors.
+func errHandler() {
```
エラーハンドラーはエラーを受け取って処理するもの（ハンドリングするもの)という意味合いで使われることが多いので、ここではvalidateとかが関数名として適正かとおもいます

MizushimaToshihiko replied:
ありがとうございます。validateに変更させていただきました。

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+	"image"
+	"image/gif"
+	"image/jpeg"
+	_ "image/jpeg"
```
おそらく不要かとおもいます

MizushimaToshihiko replied:
気付かず残してしまいました。ありがとうございます。

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+	files, err := glob(p.Path, p.PreFormat)
+
+	if err != nil {
+		log.Fatal("could not find the file path.")
```
エラーに関しては全て returnして呼び出し元で処理するのが標準的な書き方なのでそのようにした方が良いです。
また元々の書き方だとerrがなんであれ、指定されたメッセージになってしまうのでその点でもerrをそのまま返した方が良いです。

```
func (p *PicConverter) Conv() {
　　 files, err := glob(p.Path, p.PreFormat)
	if err != nil {
		log.Fatal("could not find the file path.")
        }
```
↓
```
func (p *PicConverter) Conv() error {
　　 files, err := glob(p.Path, p.PreFormat)
	if err != nil {
		return err
        }
```

MizushimaToshihiko replied:
ありがとうございます。

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+		f, err := os.Open(file)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, " the file could not open.")
+			os.Exit(1)
```
上のエラーの指摘と同じようなことですが、ライブラリ内でos.Exitを呼ばない方が良いです。コードがcli上で動くことを前提としてしまうため、望ましくないです

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+		output, err := os.Create(baseName(file) +
+			"_converted" +
+			"." +
+			p.AfterFormat)
```
+で結合しても良いですがfmt.Sprintfを使っても読みやすそうです
fmt.Sprintf("%v_converted.%v",baseName(file),p.AfterFormat)

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+	var err error
+	for _, f := range format {
+		err = filepath.Walk(path,
+			func(path string, info os.FileInfo, err error) error {
```
filepath.Walkでは抽出されるものとしてディレクトリも含んでしまうので、ディレクトリの判別処理が必要です

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// NewPicConverter is the constructor for a PicConverter.
+func NewPicConverter(Path string, PreFormat string, AfterFormat string) *PicConverter {
```
goの場合関数の引数はキャメルケース（最初が小文字である）ことの方が自然なので、それに揃えた方が良いです

MizushimaToshihiko replied:
すみません。この関数はmain.goのmain関数にエクスポートするような書き方をしたので、そのままにさせていただきました。

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+package picconvert
```
ディレクトリ名はpackageと合わせた方が自然なので、pic-convはpicconvertで良いかとおもいます

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+
+// PicConverter is User-defined type for converting
+// a picture file format has root path, pre-conversion format,
+// and after-conversion format.
```
全般的にコメントが手厚く書いてあって良いかと思います！

MizushimaToshihiko replied:
ありがとうございます！

gosagawa commented on kadai1/Mizushima/test.sh:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
```
次の課題以降でGoでのテストの書き方を学びますが、取り急ぎシェルでもテストが書いてあるのは素晴らしいと思います！

MizushimaToshihiko replied:
ありがとうございます。テストを書くのは割と楽しいですよね？（俺だけ？）。

gosagawa commented on kadai1/Mizushima/pic-conv/picconvert.go:
```diff
@@ -0,0 +1,118 @@
+			os.Exit(1)
+		}
+
+		// converting the file.
```
switchを使っても綺麗に描けるかと思います
```
switch p.AfterFormat {
     case "png":
         err = png.Encode(output, img)         
     case "jpg","jpeg":
         err = jpeg.Encode(output, img, nil)
     case "gif":
         err = gif.Encode(output, img, nil)
     default:
         // error
}
```

MizushimaToshihiko replied:
switch-caseの方が綺麗ですね。ありがとうございます。

gosagawa commented on kadai1/Mizushima/.gitignore:
```diff
@@ -0,0 +1,4 @@
+*.exe
+test/*_converted.*
```
テスト用のデータはtestdataというディレクトリ名にすると良いです。testdataというディレクトをpackage名としてみなさないためです。

MizushimaToshihiko replied:
Goのtestは前にどこかで見たときによく理解できなかったので、次のステップでしっかり勉強します。

gosagawa commented on kadai1/Mizushima/.gitignore:
```diff
@@ -0,0 +1,4 @@
+*.exe
```
おそらくexeファイルは作られないかと思うので不要かと思います。一般的にはMakefileで特定のディレクトい(bin/とか)にバイナリを出して、binごとgitignoreすることが多いです

MizushimaToshihiko replied:
ありがとうございます。

MizushimaToshihiko committed:
頂いたレビューを元に修正
※ gosagawaさんレビューありがとうございました。

picconvert.go
 _ "image/jpeg" 削除
 Conv()関数
 - errorを返すように変更
 - //converting the file の所でswitch文を使うよう変更
 glob関数
 - filepath.Walkのif文で！info.isDir()を追加（ディレクトリを除く為）

main.go
 関数名変更 isCorrespond ⇒  isSupportedFormat, errHandler ⇒ validate
 main関数
 - Conv関数から返ってきたerrorの処理を追加

その他
 - コメントのアヤシイ英語を修正（さらに怪しくなったかも）
 - ディレクトリ名変更 test ⇒  testdata, pic-conv ⇒  picconvert
 - go.mod 使用していないライブラリの記述を削除
 - test.sh テスト失敗した時の表示追加

MizushimaToshihiko commented:
@gosagawa さん
ありがとうござました。ご指摘いただきました部分を訂正いたしました。
もしよろしければ再レビューいただけますと幸いです。

MizushimaToshihiko committed:
kadai1から各ファイルコピー、picconvertのテストを追加

MizushimaToshihiko committed:
picconvert_test.go修正

MizushimaToshihiko committed:
ブランチを間違えました

MizushimaToshihiko committed:
go.modとmain.goの自作パッケージのimportを修正

MizushimaToshihiko commented:
後続の方が増えてきたのでCloseします。
